zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/zeppelinhub/websocket/ZeppelinClient.java

```diff
@@ -0,0 +1,264 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * Zeppelin websocket client.
+ *
+ */
+public class ZeppelinClient {
+  private static final Logger LOG = LoggerFactory.getLogger(ZeppelinClient.class);
+  private final URI zeppelinWebsocketUrl;
+  private final String zeppelinhubToken;
+  private final WebSocketClient wsClient;
+  private static Gson gson;
+  private ConcurrentHashMap<String, Session> zeppelinConnectionMap;
+  private static ZeppelinClient instance = null;
+
+  private Authentication authModule;
+
+  public static ZeppelinClient initialize(String zeppelinUrl, String token, 
+      ZeppelinConfiguration conf) {
+    if (instance == null) {
+      instance = new ZeppelinClient(zeppelinUrl, token, conf);
+    }
+    return instance;
+  }
+
+  public static ZeppelinClient getInstance() {
+    return instance;
+  }
+
+  private ZeppelinClient(String zeppelinUrl, String token, ZeppelinConfiguration conf) {
+    zeppelinWebsocketUrl = URI.create(zeppelinUrl);
+    zeppelinhubToken = token;
+    wsClient = createNewWebsocketClient();
+    gson = new Gson();
+    zeppelinConnectionMap = new ConcurrentHashMap<>();
+    authModule = Authentication.initialize(token, conf);
+    if (authModule != null) {
+      SchedulerService.getInstance().addOnce(authModule, 10);
+    }
+    LOG.info("Initialized Zeppelin websocket client on {}", zeppelinWebsocketUrl);
+  }
+
+  private WebSocketClient createNewWebsocketClient() {
+    SslContextFactory sslContextFactory = new SslContextFactory();
+    WebSocketClient client = new WebSocketClient(sslContextFactory);
+    //TODO(khalid): other client settings
+    return client;
+  }
+
+  public void start() {
+    try {
+      if (wsClient != null) {
+        wsClient.start();
+      } else {
+        LOG.warn("Cannot start zeppelin websocket client - isn't initialized");
+      }
+    } catch (Exception e) {
+      LOG.error("Cannot start Zeppelin websocket client", e);
+    }
+  }
+
+  public void stop() {
+    try {
+      if (wsClient != null) {
+        removeAllZeppelinConnections();
+        wsClient.stop();
+      } else {
+        LOG.warn("Cannot stop zeppelin websocket client - isn't initialized");
+      }
+    } catch (Exception e) {
+      LOG.error("Cannot stop Zeppelin websocket client", e);
+    }
+  }
+
+  public String serialize(Message zeppelinMsg) {
+    if (credentialsAvailable()) {
+      zeppelinMsg.principal = authModule.getPrincipal();
+      zeppelinMsg.ticket = authModule.getTicket();
+      zeppelinMsg.roles = authModule.getRoles();
+    }
+    String msg = gson.toJson(zeppelinMsg);
+    return msg;
+  }
+
+  private boolean credentialsAvailable() {
+    return Authentication.getInstance() != null && Authentication.getInstance().isAuthenticated();
+  }
+
+  public Message deserialize(String zeppelinMessage) {
+    if (StringUtils.isBlank(zeppelinMessage)) {
+      return null;
```

Comment:
That would be the case if `Message` class had `EMPTY` element. However, currently `null` used across the code-base to do all the checks for instances of `Message`, so I decided to keep consistent with it.